File: TimeSeriesSRC/basefunctions/sdiff.py
from . import makerow
import numpy as np

def func_sdiff (y,d,p=1) :
	"""Apply seasonal (or regular) differencing to a time series.

    Computes the ``d``-th order difference at period ``p``:

    .. math::

        \\nabla_p^d\\, y(t) = \\nabla_p^{d-1}\\, y(t) - \\nabla_p^{d-1}\\, y(t-p)

    For ``p = 1`` this reduces to ordinary differencing (``numpy.diff``).

    Parameters
    ----------
    y : array-like
        1-D or row-vector input series.
    d : int
        Number of differences to apply. ``d = 0`` returns ``y`` unchanged.
    p : int, optional
        Seasonal period. Default 1 (ordinary differencing).

    Returns
    -------
    yd : ndarray, shape (1, N - d*p)
        Differenced series.  Each application of the operator reduces the
        length by ``p``.

    Raises
    ------
    Exception
        If ``d * p`` is larger than or equal to the length of ``y``.

    Examples
    --------
    >>> import numpy as np
    >>> from TimeSeriesSRC.basefunctions.sdiff import func_sdiff
    >>> y = np.arange(1.0, 21.0)
    >>> yd = func_sdiff(y, d=1, p=4)   # seasonal difference at lag 4
    >>> yd.shape
    (1, 16)

    See Also
    --------
    uniAnal : Passes ``diff`` / ``per`` arguments directly to this function.
	"""


	# Make y into row format

	y = makerow.func_makerow(y)
	m, num_pts = y.shape

	if num_pts <= (d * p):
		raise Exception('d*p is larger than the number of points.')

	if p == 1:
		yd = np.diff(y, d)
	else:
		yd = y

		for i in range(d):
			yd = yd[:,  p: num_pts] - yd[:, : num_pts - p]
			num_pts = num_pts - p

	return yd

File: Matlab_Files/PredictMod/sdiff.m
function [yd] = sdiff(y,d,p)
%SDIFF Seasonal differencing
%	
%	Syntax
%
%	  [yd] = sdiff(y,d,p)
%
%	Description
%	
%	  SDIFF differences the data a specified number of 
%	  times.  Each difference involves subtracting each 
%	  original time point from the point one period away.
%     
%	
%	  SDIFF(Y,D,P) takes these inputs,
%	    Y - 1xN vector containing the input sequence.
%	    D - Number of differences to take.
%	    P - Period of the difference; default = 1.
%	  and returns,
%	    YD  - RxQ vector containing the differenced series
%		
%	Examples
%
%	  Here is how to difference a sequence twice
%	  at a period of 4.
%	
%	    y=[1:20].^2;
%	    yd=sdiff(y,2,4);
%
%	Algorithm
%
%	    yd = yd(1+i*p:num_pts) - yd(1:num_pts-i*p);
%
%	See also DIFF.

% Yong Hu, Martin Hagan, 9-15-00
% $Revision: 1.0 $ $Date: 21-Sep-2000 14:37:36 $


if nargin < 2
  error('Wrong number of arguments.');
end
if nargin <3
  p=1;
end

% Make y into row format
y = makerow(y);
[m,num_pts] = size(y);

if num_pts<=(d*p),
  error('d*p is larger than the number of points.');
end

if p==1
  yd = diff(y,d,2);
else
  yd = y;
  for i=1:d,
    yd = yd(:,1+p:num_pts) - yd(:,1:num_pts-p);
    num_pts = num_pts-p;
  end
end
